src/ts/pklLspDistribution.ts

```diff
@@ -0,0 +1,171 @@
+// limitations under the License.
+//===----------------------------------------------------------------------===//
+
+import path from "node:path";
+import config from "./config";
+import { getJavaDistribution } from "./javaDistribution";
+import { debounce, exec, isRegularFile } from "./utils";
+import Semver from "./Semver";
+import * as vscode from "vscode";
+import fs from "fs/promises";
+import {
+  BUNDLED_LSP_VERSION,
+  COMMAND_OPEN_WORKSPACE_SETTINGS,
+  CONFIG_LSP_PATH,
+  LSP_DISTRIBUTIONS_DIR,
+} from "./consts";
+import logger from "./clients/logger";
+
+const emitter = new vscode.EventEmitter<LspDistribution>();
+
+export const onDidChangeLspDistribution = emitter.event;
+
+export let currentLspDistribution: LspDistribution | undefined = undefined;
+
+export type LspDistribution = {
+  path: string;
+  version: Semver;
+};
+
+export const getLspDistribution = (): Promise<LspDistribution> => {
+  return new Promise((resolve) => {
+    if (currentLspDistribution !== undefined) {
+      resolve(currentLspDistribution);
+      return;
+    }
+    const disposables: vscode.Disposable[] = [];
+    onDidChangeLspDistribution(
+      (distribution) => {
+        resolve(distribution);
+        disposables.every((it) => it.dispose());
+      },
+      null,
+      disposables
+    );
+  });
+};
+
+export const bundledDistribution: LspDistribution = {
+  path: path.join(__dirname, "pkl-lsp.jar"),
+  version: Semver.parse(BUNDLED_LSP_VERSION)!!,
+};
+
+const getLspVersion = async (jarPath: string): Promise<Semver | undefined> => {
+  const javaDistribution = await getJavaDistribution();
+  const { stdout } = await exec(`${javaDistribution.path} -jar ${jarPath} --version`);
+  const versionStr = stdout
+    .replace(/\r?\n$/, "")
+    .split(" version ")
+    .findLast((it) => true);
+  if (versionStr === undefined) {
+    logger.log(
+      `Got malformed version output from jar file at ${jarPath}: ${stdout}. Expected "pkl-lsp version <version>"`
+    );
+    return;
+  }
+  const semver = Semver.parse(versionStr);
+  if (semver === undefined) {
+    logger.log(`Got malformed semver string from jar file at ${jarPath}: ${versionStr}`);
+    return;
+  }
+  return semver;
+};
+
+const CTA_CONFIGURE_LSP_PATH = "Configure path to pkl-lsp";
+
+const tellInvalidConfiguredLspPath = async () => {
+  const response = await vscode.window.showWarningMessage(
+    `Configured path ${config.lspPath} is not a valid lsp jar.`,
+    CTA_CONFIGURE_LSP_PATH
+  );
+  if (response === CTA_CONFIGURE_LSP_PATH) {
+    vscode.commands.executeCommand(COMMAND_OPEN_WORKSPACE_SETTINGS, CONFIG_LSP_PATH);
+  }
+};
+
+const handleConfiguredLspDistribution = async (lspPath: string) => {
+  try {
+    const version = await getLspVersion(lspPath);
+    if (version === undefined) {
+      tellInvalidConfiguredLspPath();
+      return;
+    }
+    // permit a higher version if it exists, but warn users about it.
```

Comment:
nit: not only higher, it may also be an older version.

Reply:
Yeah, good point!